eureka-client/src/main/java/com/netflix/discovery/InstanceInfoReplicator.java

```diff
@@ -0,0 +1,114 @@
+ * - update tasks can be scheduled on-demand via onDemandUpdate()
+ * - task processing is rate limited by burstSize
+ * - a new update task is always scheduled automatically after an earlier update task. However if an on-demand task
+ *   is started, the scheduled automatic update task is discarded (and a new one will be scheduled after the new
+ *   on-demand update).
+ *
+ *   @author dliu
+ */
+class InstanceInfoReplicator implements Runnable {
+    private static final Logger logger = LoggerFactory.getLogger(InstanceInfoReplicator.class);
+
+    private final DiscoveryClient discoveryClient;
+    private final InstanceInfo instanceInfo;
+
+    private final int replicationIntervalSeconds;
+    private final ScheduledExecutorService scheduler;
+    private final AtomicReference<Future> scheduledPeriodicRef;
+
+    private final AtomicBoolean started;
+    private final RateLimiter rateLimiter;
+    private final int burstSize;
+    private final int allowedRatePerMinute;
+
+    InstanceInfoReplicator(DiscoveryClient discoveryClient, InstanceInfo instanceInfo, int replicationIntervalSeconds, int burstSize) {
+        this.discoveryClient = discoveryClient;
+        this.instanceInfo = instanceInfo;
+        this.scheduler = Executors.newScheduledThreadPool(1,
+                new ThreadFactoryBuilder()
+                        .setNameFormat("DiscoveryClient-InstanceInfoReplicator-%d")
+                        .setDaemon(true)
+                        .build());
+
+        this.scheduledPeriodicRef = new AtomicReference<Future>();
+
+        this.started = new AtomicBoolean(false);
+        this.rateLimiter = new RateLimiter(TimeUnit.MINUTES);
+        this.replicationIntervalSeconds = replicationIntervalSeconds;
+        this.burstSize = burstSize;
+
+        this.allowedRatePerMinute = 60 * this.burstSize / this.replicationIntervalSeconds;
+        logger.info("InstanceInfoReplicator onDemand update allowed rate per min is {}", allowedRatePerMinute);
+    }
+
+    public void start(int initialDelayMs) {
+        if (started.compareAndSet(false, true)) {
+            instanceInfo.setIsDirty();  // for initial register
+            scheduler.schedule(this, initialDelayMs, TimeUnit.SECONDS);
+        }
+    }
+
+    public void stop() {
+        scheduler.shutdown();
+        started.set(false);
+    }
+
+    public boolean onDemandUpdate() {
+        if (rateLimiter.acquire(burstSize, allowedRatePerMinute)) {
+            scheduler.submit(new Runnable() {
+                @Override
+                public void run() {
+                    logger.debug("Executing on-demand update of local InstanceInfo");
+
+                    // cancel the latest scheduled update, it will be rescheduled at the end of run()
+                    Future latestPeriodic = scheduledPeriodicRef.get();
+                    if (latestPeriodic != null && !latestPeriodic.isDone()) {
```

Comment:
May be having the last onDemand run time stored and shared between onDemand & scheduled tasks, can remove this schedule-cancel-schedule song & dance :)
In that case, the scheduled task can just complete, if the onDemand update has run within X amount of time.